pkg/odp/event_manager.go

```diff
@@ -0,0 +1,343 @@
+	eventAPIManager EventAPIManager
+	odpConfig       Config
+	processing      *semaphore.Weighted
+	logger          logging.OptimizelyLogProducer
+}
+
+// WithBatchSize sets the batch size as a config option to be passed into the NewBatchEventManager method
+func WithBatchSize(bsize int) EMOptionConfig {
+	return func(bm *BatchEventManager) {
+		bm.batchSize = bsize
+	}
+}
+
+// WithQueueSize sets the queue size as a config option to be passed into the NewBatchEventManager method
+func WithQueueSize(qsize int) EMOptionConfig {
+	return func(bm *BatchEventManager) {
+		bm.maxQueueSize = qsize
+	}
+}
+
+// WithFlushInterval sets the flush interval as a config option to be passed into the NewBatchEventManager method
+func WithFlushInterval(flushInterval time.Duration) EMOptionConfig {
+	return func(bm *BatchEventManager) {
+		bm.flushInterval = flushInterval
+	}
+}
+
+// WithQueue sets the Queue as a config option to be passed into the NewBatchEventManager method
+func WithQueue(q event.Queue) EMOptionConfig {
+	return func(bm *BatchEventManager) {
+		bm.eventQueue = q
+	}
+}
+
+// WithSDKKey sets the SDKKey used for logging
+func WithSDKKey(sdkKey string) EMOptionConfig {
+	return func(bm *BatchEventManager) {
+		bm.sdkKey = sdkKey
+	}
+}
+
+// WithEventAPIManager sets eventAPIManager as a config option to be passed into the NewBatchEventManager method
+func WithEventAPIManager(eventAPIManager EventAPIManager) EMOptionConfig {
+	return func(bm *BatchEventManager) {
+		bm.eventAPIManager = eventAPIManager
+	}
+}
+
+// WithConfig sets odpConfig option to be passed into the NewBatchEventManager method
+func WithConfig(odpConfig Config) EMOptionConfig {
+	return func(bm *BatchEventManager) {
+		bm.odpConfig = odpConfig
+	}
+}
+
+// NewBatchEventManager returns a new instance of BatchEventManager with options
+func NewBatchEventManager(options ...EMOptionConfig) *BatchEventManager {
+	bm := &BatchEventManager{processing: semaphore.NewWeighted(int64(maxFlushWorkers))}
+
+	for _, opt := range options {
+		opt(bm)
+	}
+
+	bm.logger = logging.GetLogger(bm.sdkKey, "BatchEventManager")
+
+	if bm.maxQueueSize == 0 {
+		bm.maxQueueSize = DefaultEventQueueSize
+	}
+
+	if bm.flushInterval == 0 {
+		bm.flushInterval = DefaultEventFlushInterval
+	}
+
+	if bm.batchSize == 0 {
+		bm.batchSize = DefaultBatchSize
+	}
+
+	if bm.batchSize > bm.maxQueueSize {
+		bm.logger.Warning(
+			fmt.Sprintf("Batch size %d is larger than queue size %d.  Setting to defaults",
+				bm.batchSize, bm.maxQueueSize))
+
+		bm.batchSize = DefaultBatchSize
+		bm.maxQueueSize = DefaultEventQueueSize
+	}
+
+	if bm.eventQueue == nil {
+		bm.eventQueue = event.NewInMemoryQueueWithLogger(bm.maxQueueSize, bm.logger)
+	}
+
+	if bm.eventAPIManager == nil {
+		bm.eventAPIManager = NewEventAPIManager(bm.sdkKey, nil)
+	}
+
+	return bm
+}
+
+// Start does not do any initialization, just starts the ticker
+func (bm *BatchEventManager) Start(ctx context.Context) {
+	if !bm.IsOdpServiceIntegrated() {
```

Comment:
I think this should be ready? please check in other sdks.

Reply:
This is similar to `odpServiceIntegrated` property in `swift-sdk`